src/gc/proxy.d

```diff
@@ -62,57 +68,158 @@ private
+        pthis.gc_calloc = function void*(size_t sz, uint ba, const TypeInfo ti) { return _mallocGC.calloc( sz);};
+        pthis.gc_realloc = function void*(void* p, size_t sz, uint ba, const TypeInfo ti) { return _mallocGC.realloc( p, sz);};
+        pthis.gc_extend = function size_t(void* p, size_t mx, size_t sz, const TypeInfo ti) { return 0;};
+        pthis.gc_reserve = function size_t(size_t sz) { return 0;};
+        pthis.gc_free = function void(void* p) { _mallocGC.free(p);};
+
+        pthis.gc_addrOf = function void*(void* p) { return null;};
+        pthis.gc_sizeOf = function size_t(void* p) { return 0;};
 
-        pthis.gc_getAttr = &gc_getAttr;
-        pthis.gc_setAttr = &gc_setAttr;
-        pthis.gc_clrAttr = &gc_clrAttr;
+        pthis.gc_query = function BlkInfo(void* p) { return BlkInfo.init;};
+        pthis.gc_stats = function GCStats() { return GCStats.init;};
 
-        pthis.gc_malloc = &gc_malloc;
-        pthis.gc_qalloc = &gc_qalloc;
-        pthis.gc_calloc = &gc_calloc;
-        pthis.gc_realloc = &gc_realloc;
-        pthis.gc_extend = &gc_extend;
-        pthis.gc_reserve = &gc_reserve;
-        pthis.gc_free = &gc_free;
+        pthis.gc_addRoot = function void(void* p) { _mallocGC.addRoot(p);};
+        pthis.gc_addRange = function void(void* p, size_t sz, const TypeInfo ti) { _mallocGC.addRange( p, sz, ti );};
 
-        pthis.gc_addrOf = &gc_addrOf;
-        pthis.gc_sizeOf = &gc_sizeOf;
+        pthis.gc_removeRoot = function void(void* p) { _mallocGC.removeRoot(p);};
+        pthis.gc_removeRange = function void(void*p) { _mallocGC.removeRange(p);};
+        pthis.gc_runFinalizers = function void(in void[] segment) { };
 
-        pthis.gc_query = &gc_query;
+        pthis.gc_inFinalizer = function bool() { return false;};
 
-        pthis.gc_addRoot = &gc_addRoot;
-        pthis.gc_addRange = &gc_addRange;
+        pthis.gc_setProxy = function void(Proxy* p) {
+            foreach( r; _mallocGC.roots[0 .. _mallocGC.nroots] )
+                p.gc_addRoot( r );
 
-        pthis.gc_removeRoot = &gc_removeRoot;
-        pthis.gc_removeRange = &gc_removeRange;
-        pthis.gc_runFinalizers = &gc_runFinalizers;
+            foreach( r; _mallocGC.ranges[0 .. _mallocGC.nranges] )
+                p.gc_addRange( r.pos, r.len, r.ti );
+        };
+        pthis.gc_clrProxy= function void(){
+            foreach( r; _mallocGC.ranges[0 .. _mallocGC.nranges] )
+                proxy.gc_removeRange( r.pos );
 
-        pthis.gc_inFinalizer = &gc_inFinalizer;
+            foreach( r; _mallocGC.roots[0 .. _mallocGC.nroots] )
+                proxy.gc_removeRoot( r );
+        };
     }
+
 }
 
 extern (C)
 {
-
     void gc_init()
     {
-        _gc.initialize();
+
+        config.initialize();
+
+        if (config.malloc)
+        {
+            initMallocGCProxy();
+        }
+        else
+        {
+            _conservativeGC.initialize();
+            initConservativeGCProxy();
+        }
```

Comment:
Always call the init functions of all GCs and let them do the right thing based on the config.
```
config.initialize();
conservative.instance.initialize(); // instance being the singleton
malloc.instance.initialize();
```